hooks/matrix.ts

```diff
@@ -0,0 +1,295 @@
+
+  const setupStarted = useRef<boolean>(false);
+
+  /**
+   * Creates a color scale array map used to be passed to components where the key is the index
+   */
+  const createColorScaleArrayMap = useCallback(
+    (data: any) => {
+      // choose theme color based on state provided
+      const [, , { data: zData }] = data;
+      let colorScale: Array<Array<number>>;
+      if (colorThemeOption.value === 'parula') {
+        colorScale = generateScale(parula);
+      } else if (colorThemeOption.value === 'viridis') {
+        colorScale = generateScale(viridis);
+      } else {
+        colorScale = [];
+      }
+
+      // create resulting color map
+      const result = zData.map((_z: any, i: number) => {
+        if (colorThemeOption.value !== 'parula' && colorThemeOption.value !== 'viridis') {
+          return generateRainbowColor(zData.length, i);
+        } else {
+          return getColor(colorScale, zData.length, i);
+        }
+      });
+      return result;
+    },
+    [colorThemeOption.value]
+  );
+
+  // Side Effect for Updating Data
+  useEffect(() => {
+    const fetchMatrixData = async (config: any) => {
+      const xPromise = ConfigService.http.get(config.xURL);
+      const yPromise = ConfigService.http.get(config.yURL);
+      const zPromise = ConfigService.http.get(config.zURL);
+      const xyzPromise = ConfigService.http.get(config.xysURL);
+
+      // Batch the requests in Promise.all so they can run in parallel:
+      const data = await Promise.all([xPromise, yPromise, zPromise, xyzPromise]);
+      const parsedData = parseMatrixData(config, data);
+
+      // Set state after the request completes
+      setData(data);
+      setMatrixData(parsedData);
+      setStyles(config.styles);
+    };
+
+    if (setupStarted.current) return;
+    setupStarted.current = true;
+
+    try {
+      const config = getConfigObject(matrixConfigs);
+      fetchMatrixData(config);
+    } catch (error: any) {
+      dispatchError({ error });
+    }
+  }, [matrixConfigs, dispatchError, createColorScaleArrayMap]);
+
+  // Side Effect for Updating Color Scale
+  useEffect(() => {
+    if (data && Array.isArray(data)) {
+      const newColorScaleMap = createColorScaleArrayMap(data);
+      setColorScaleMap(newColorScaleMap);
+    }
+  }, [data, colorThemeOption.value, createColorScaleArrayMap]);
+
+  return {
+    dispatchError,
+    errors,
+    matrixData,
+    setMatrixData,
+    setupStarted,
+    colorOptions,
+    colorThemeOption,
+    setColorThemeOption,
+    colorScaleMap,
+    styles,
+  };
+};
+
+/**
+ * Generate a link for combination of x and y filter on the main table
+ * @param  {object} x the x object, must have .title and .id
+ * @param  {object} y the y object, must have .title and .id
+ * @return {string} returns the url string to be used
+ */
+const generateLink = (config: any, x?: any, y?: any, z?: any) => {
```

Comment:
Why is this function down here, and the other "util" functions above the hook? Either move this to be above the `useMatrix`, or move the other function to be after this function.